### PR TITLE
CORDA-1724 Add tx access right validity check during resolution

### PR DESCRIPTION
**Pull request by tudor-malene**
see https://r3-cev.atlassian.net/browse/CORDA-1724

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt [hunk 1]
```diff
@@ -42,6 +44,25 @@ open class DataVendingFlow(val otherSideSession: FlowSession, val payload: Any)
     override fun call(): Void? {
         // The first payload will be the transaction data, subsequent payload will be the transaction/attachment data.
         var payload = payload
+
+        // Depending on who called this flow, the type of the initial payload is different.
+        // The authorisation logic is to maintain a dynamic list of transactions that the caller is authorised to make based on the transactions that were made already.
+        // Each time an authorised transaction is requested, the input transactions are added to the list.
+        // Todo: should we remove a txId from the list once it has been requested? This would keep the list smaller, but it would fail if the requester asks for the same tx twice. Is that desired?
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt]: The other side shouldn't request the same tx twice, we can easily make that a protocol violation and throw an error if it happens.

File: core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt [hunk 2]
```diff
@@ -56,14 +77,44 @@ open class DataVendingFlow(val otherSideSession: FlowSession, val payload: Any)
...
+                            ?: throw FetchDataFlow.HashNotFound(it)
                 }
             }
         }
     }
+
+    @Suspendable
+    private fun getInputTransactions(tx: SignedTransaction): Set<SecureHash> = tx.inputs.map { it.txhash }.toSet()
+
+    class TransactionAuthorisationFilter(val validRequests: MutableSet<SecureHash> = mutableSetOf(), val acceptAll: Boolean = false) {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt]: private class?

File: core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt [hunk 3]
```diff
@@ -56,14 +77,44 @@ open class DataVendingFlow(val otherSideSession: FlowSession, val payload: Any)
...
+        @Suspendable
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt]: These don't need to be suspendable, right?

File: core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt [hunk 4]
```diff
@@ -56,14 +77,44 @@ open class DataVendingFlow(val otherSideSession: FlowSession, val payload: Any)
...
+        fun addAuthorised(txs: Set<SecureHash>): TransactionAuthorisationFilter {
+            validRequests.addAll(txs)
+            return this
+        }
+    }
 }
+
+/**
+ * This is a wildcard payload to be used by the invoker of the [DataVendingFlow] to allow unlimited access to it's vault.
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt]: typo: it's ... but not sure why it's here or what that's for. Is this feature really needed?

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt [hunk 5]
```diff
@@ -49,6 +49,8 @@ sealed class FetchDataFlow<T : NamedByHash, in W : Any>(
 
     class HashNotFound(val requested: SecureHash) : FlowException()
 
+    class IllegalTransactionRequest(val requested: SecureHash) : FlowException("Illegally trying to access transaction: ${requested}")
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: Better error message might be "Illegal attempt to request a transaction that is not in the transitive dependency graph of the sent transaction"

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt [hunk 6]
```diff
@@ -64,7 +66,7 @@ sealed class FetchDataFlow<T : NamedByHash, in W : Any>(
     }
 
     @Suspendable
-    @Throws(HashNotFound::class)
+    @Throws(HashNotFound::class, IllegalTransactionRequest::class)
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: This is internal code, so we don't need throws at all here I guess (only used from Kotlin).

File: core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt (hunk 4 above)
Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/flows/SendTransactionFlow.kt]: It is currently for some tests we have.

**Comment by tudor-malene** [comment]
Back to you @mikehearn

**Pull request merged**